client.go

```diff
@@ -53,173 +52,28 @@ func New(resolver kafka.NameResolver, logger *zap.Logger, scope tally.Scope) *Cl
 		resolver: resolver,
 		logger:   logger,
 		tally:    scope,
-		saramaSyncProducerConstructor: newSyncProducer,
+		saramaSyncProducerConstructor: consumer.NewSaramaProducer,
 		saramaConsumerConstructor:     consumer.NewSaramaConsumer,
+		clusterConsumerConstructor:    consumer.NewClusterConsumer,
 	}
 }
 
-// NewConsumer returns a new instance of kafka consumer
-func (c *Client) NewConsumer(config *kafka.ConsumerConfig, options ...ConsumerOption) (kafka.Consumer, error) {
-	var err error
-	topicList := config.TopicList
-	if err = validateTopicListFromSingleCluster(topicList); err != nil {
-		return nil, err
-	}
-	clusterName := topicList[0].Cluster
-
+// NewConsumer returns a new instance of kafka consumer.
+//
+// It is possible for NewConsumer to start a consumer which consumes from a subset of topics.
+// If partial construction has occurred, an error will be returned and you can get a list of topics that are not
+// being consumed by using the HasConsumerBuildError() function.
+func (c *Client) NewConsumer(config *kafka.ConsumerConfig) (kafka.Consumer, error) {
```

Comment:
This API is still confusing to me. If this method returns error, what does that mean ? Does it mean partial construction or no construction at all ? How do I know to call HasConsumerBuildError() if I don't read the docs . The common go idiom is to return immediately if err != nil. However, in this case, that's not the right thing to do (causes resource leak).

I am not a fan of partial construction as you already know. But since we are going down this route, lets go down the route of least surprise for Golang developers i.e. when partial construction succeeds, lets not return an error at all from this function. Instead, let's just document that partial construction is possible and change the method signature a bit to make it explicit i.e. New(config) (kafka.Consumer, consumedTopics []string, error) - let's also document that callers can check if consumedTopics != requestedTopics if they care about it.

As for getting visibility into the errors, let's just log the errors / emit metrics.  

In addition to the above, as we discussed earlier, we should add an Option/Flag in ConsumerConfig to enable/disable this behavior. And the default must be to disable partial construction.

Reply:
I have set an optional flag for enabling partial construction. If partial construction is enabled, it will return the partial consumer and a generic `error`. The user may use `HasConsumerBuildError` to recover structured error messages.